provider-proxy/src/lib.rs

```diff
@@ -0,0 +1,251 @@
+                                    err = e.as_ref() as &dyn std::error::Error,
+                                    "Failed to save cache body"
+                                );
+                            }
+                        });
+                    }),
+                )
+            });
+
+            Ok(body_collector.map(|b| BoxBody::new(b)))
+        } else {
+            tracing::warn!("Skipping caching of non-success response: {:?}", response);
+            Ok(response)
+        }
+    }
+}
+
+#[derive(Parser, Debug)]
+#[command(version, about, long_about = None)]
+pub struct Args {
+    /// Path to the cache directory
+    #[arg(long, default_value = "request_cache")]
+    pub cache_path: PathBuf,
+    /// Port to listen on
+    #[arg(long, default_value = "3003")]
+    pub port: u16,
+}
+
+pub async fn run_server(args: Args, server_started: oneshot::Sender<SocketAddr>) {
+    use tracing_subscriber::EnvFilter;
+
+    tracing_subscriber::fmt()
+        .with_env_filter(
+            EnvFilter::builder()
+                .with_default_directive(LevelFilter::INFO.into())
+                .from_env_lossy(),
+        )
+        .try_init()
+        .unwrap();
+
+    std::fs::create_dir_all(&args.cache_path).expect("Failed to create cache directory");
+
+    rustls::crypto::ring::default_provider()
+        .install_default()
+        .expect("Failed to install rustls ring provider");
+
+    let root_cert = make_root_cert();
+
+    let proxy = MitmProxy::new(
+        // This is the root cert that will be used to sign the fake certificates
+        Some(root_cert),
+        Some(Cache::new(128)),
+    );
+
+    let client = reqwest::Client::new();
+    let (server_addr, server) = proxy
+        .bind(
+            ("127.0.0.1", args.port),
+            service_fn(move |req: hyper::Request<hyper::body::Incoming>| {
+                let client = client.clone();
+                let cache_path = args.cache_path.clone();
+                async move {
+                    let (parts, body) = req.into_parts();
+                    let body_bytes = body
+                        .collect()
+                        .await
+                        .with_context(|| "Failed to collect body")?
+                        .to_bytes();
+                    let bytes_request = hyper::Request::from_parts(parts, body_bytes);
+                    let response = check_cache(cache_path, &bytes_request, || async {
+                        let request: reqwest::Request =
+                            bytes_request.clone().try_into().with_context(|| {
+                                "Failed to convert Request from `hyper` to `reqwest`"
+                            })?;
+                        Ok(http::Response::from(client.execute(request).await?).map(BoxBody::new))
+                    })
+                    .await?;
+
+                    Ok::<_, anyhow::Error>(response)
+                }
+            }),
```

Comment:
we need to set the `x-tensorzero-cache` header here and also test that it gets set in the tests

Reply:
if a cache hit oops so probably needs to be somewhere else

Reply:
I added a `x-tensorzero-provider-proxy-cache` header and updated the tests